### PR TITLE
Major revision of the draw API

### PR DESCRIPTION
**Pull request by emilwestergren**

### CODE REVIEW

**Comment by davidhesselbom** [comment]
If it's not too much trouble, please try rebasing this on `master` one last time.

File: source/draw/gpu/GpuSurface.ooc [hunk 1]
```diff
@@ -16,14 +16,110 @@
...
 	clear: abstract func
-	recycle: abstract func
+	draw: func ~general (action: Func) {
+		this _bind()
+		this _context setViewport(this viewport)
+		this _context enableBlend(this blend)
+		action()
+		this _unbind()
+	}
+	draw: func ~WithoutBind (destination: IntBox2D, map: GpuMap) {
```
Comment (davidhesselbom) [inline source/draw/gpu/GpuSurface.ooc]: This function causes a lot of leaks, according to valgrind. My guess is the `func` needs to be freed after use.

Comment (davidhesselbom) [inline source/draw/gpu/GpuSurface.ooc]: OK, so, as we agreed, fix the memory leaks by giving the closure its own variable and free it after use.

Comment (emilwestergren) [inline source/draw/gpu/GpuSurface.ooc]: Fixed.

**Comment by emilwestergren** [comment]
Updated and rebased on upstream/master.

**Pull request merged**